src/System.Private.CoreLib/shared/System/IO/Stream.cs

```diff
@@ -189,6 +189,65 @@ private int GetCopyBufferSize()
             return bufferSize;
         }
 
+        public virtual void CopyTo(ReadOnlySpanAction<byte, object> callback, object state, int bufferSize) =>
+            CopyTo(new WriteCallbackStream(callback, state), bufferSize);
+
+        public virtual Task CopyToAsync(Func<object, ReadOnlyMemory<byte>, CancellationToken, Task> callback, object state, int bufferSize, CancellationToken cancellationToken) =>
+            CopyToAsync(new WriteCallbackStream(callback, state), bufferSize, cancellationToken);
+
+        private sealed class WriteCallbackStream : Stream
+        {
+            private readonly ReadOnlySpanAction<byte, object>? _action;
+            private readonly Func<object, ReadOnlyMemory<byte>, CancellationToken, Task>? _func;
+            private readonly object _state;
+
+            public WriteCallbackStream(ReadOnlySpanAction<byte, object> action, object state)
+            {
+                _action = action ?? throw new ArgumentNullException(nameof(action));
+                _state = state;
+            }
+
+            public WriteCallbackStream(Func<object, ReadOnlyMemory<byte>, CancellationToken, Task> func, object state)
+            {
+                _func = func ?? throw new ArgumentNullException(nameof(func));
+                _state = state;
+            }
+
+            public override void Write(byte[] buffer, int offset, int count)
+            {
+                if (_action == null) throw new NotSupportedException();
+                _action(new ReadOnlySpan<byte>(buffer, offset, count), _state);
+            }
+
+            public override void Write(ReadOnlySpan<byte> span)
+            {
+                if (_action == null) throw new NotSupportedException();
+                _action(span, _state);
+            }
+
+            public override Task WriteAsync(byte[] buffer, int offset, int length, CancellationToken cancellationToken)
+            {
+                if (_func == null) throw new NotSupportedException();
+                return _func(_state, new ReadOnlyMemory<byte>(buffer, offset, length), cancellationToken);
+            }
+
+            public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
+            {
+                if (_func == null) throw new NotSupportedException();
+                return new ValueTask(_func(_state, buffer, cancellationToken));
+            }
+
+            public override bool CanRead => false;
+            public override bool CanSeek => false;
+            public override bool CanWrite => true;
+            public override void Flush() => throw new NotSupportedException();
```

Comment:
Special nit here: should `Flush` no-op instead of throw?

Reply:
I agree. Fixed.